File: remote/src/main/java/com/kelsos/mbrc/ui/fragments/profile/GenreArtistsActivity.java
package com.kelsos.mbrc.ui.fragments.profile;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import butterknife.Bind;
import butterknife.ButterKnife;
import com.google.inject.Inject;
import com.kelsos.mbrc.R;
import com.kelsos.mbrc.adapters.ArtistAdapter;
import com.kelsos.mbrc.annotations.Queue;
import com.kelsos.mbrc.domain.Artist;
import com.kelsos.mbrc.presenters.GenreArtistsPresenter;
import com.kelsos.mbrc.ui.views.GenreArtistView;
import java.util.List;
import roboguice.RoboGuice;

public class GenreArtistsActivity extends AppCompatActivity
    implements GenreArtistView, ArtistAdapter.MenuItemSelectedListener {

  public static final String GENRE_ID = "genre_id";
  public static final String GENRE_NAME = "genre_name";
  @Bind(R.id.genre_artists_recycler) RecyclerView recyclerView;
  @Bind(R.id.toolbar) Toolbar toolbar;

  @Inject private ArtistAdapter adapter;
  @Inject private GenreArtistsPresenter presenter;
  private long genreId;

  public GenreArtistsActivity() {
    // Required empty public constructor
  }

  @Override public void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_genre_artists);
    RoboGuice.getInjector(this).injectMembers(this);
    ButterKnife.bind(this);
    presenter.bind(this);
    RecyclerView.LayoutManager manager = new LinearLayoutManager(this);
    recyclerView.setLayoutManager(manager);
    recyclerView.setAdapter(adapter);
    adapter.setMenuItemSelectedListener(this);
    final Bundle extras = getIntent().getExtras();
    genreId = 0;

    String title = "";

    if (extras != null) {
      genreId = extras.getLong(GENRE_ID, 0);
      title = extras.getString(GENRE_NAME, "");
    }

    setSupportActionBar(toolbar);
    final ActionBar actionBar = getSupportActionBar();

    if (actionBar != null) {
      actionBar.setDisplayHomeAsUpEnabled(true);
      actionBar.setDisplayShowHomeEnabled(true);
      actionBar.setTitle(title);
    }

    presenter.load(genreId);
  }

  @Override public void update(List<Artist> data) {
    adapter.updateData(data);
  }

  @Override public void onQueueSuccess() {

  }

  @Override public void onQueueFailure() {

  }

  private void openProfile(Artist artist) {
    Intent intent = new Intent(this, ArtistAlbumsActivity.class);
    intent.putExtra(ArtistAlbumsActivity.ARTIST_ID, artist.getId());
    intent.putExtra(ArtistAlbumsActivity.ARTIST_NAME, artist.getName());
    startActivity(intent);
  }

  @Override public void onMenuItemSelected(MenuItem menuItem, Artist artist) {
    int itemId = menuItem.getItemId();
    switch (itemId) {
      case R.id.popup_artist_play:
        presenter.queue(Queue.NOW, artist);
        break;
      case R.id.popup_artist_album:
        openProfile(artist);
        break;
      case R.id.popup_artist_queue_next:
        presenter.queue(Queue.NEXT, artist);
        break;
      case R.id.popup_artist_queue_last:
        presenter.queue(Queue.LAST, artist);
        break;
      case R.id.popup_artist_playlist:
        break;
    }
  }

  @Override public void onItemClicked(Artist artist) {
    openProfile(artist);
  }
}

File: app/src/main/kotlin/com/kelsos/mbrc/ui/fragments/profile/GenreArtistsActivity.kt
package com.kelsos.mbrc.ui.fragments.profile

import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.Toolbar
import android.view.MenuItem
import butterknife.BindView
import butterknife.ButterKnife
import com.kelsos.mbrc.R
import com.kelsos.mbrc.adapters.ArtistEntryAdapter
import com.kelsos.mbrc.annotations.Queue
import com.kelsos.mbrc.domain.Artist
import com.kelsos.mbrc.extensions.empty
import com.kelsos.mbrc.presenters.GenreArtistsPresenter
import com.kelsos.mbrc.ui.views.GenreArtistView
import toothpick.Toothpick
import javax.inject.Inject

class GenreArtistsActivity : AppCompatActivity(), GenreArtistView, ArtistEntryAdapter.MenuItemSelectedListener {
  @BindView(R.id.genre_artists_recycler) internal lateinit var recyclerView: RecyclerView
  @BindView(R.id.toolbar) internal lateinit var toolbar: Toolbar

  @Inject lateinit var adapter: ArtistEntryAdapter
  @Inject lateinit var presenter: GenreArtistsPresenter
  private var genreId: Long = 0

  public override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_genre_artists)
    val scope = Toothpick.openScopes(application, this)
    Toothpick.inject(this, scope)

    ButterKnife.bind(this)
    presenter.bind(this)
    val manager = LinearLayoutManager(this)
    recyclerView.layoutManager = manager
    recyclerView.adapter = adapter
    adapter.setMenuItemSelectedListener(this)
    val extras = intent.extras
    genreId = 0

    var title = String.empty

    if (extras != null) {
      genreId = extras.getLong(GENRE_ID, 0)
      title = extras.getString(GENRE_NAME, String.empty)
    }

    setSupportActionBar(toolbar)
    supportActionBar?.setDisplayHomeAsUpEnabled(true)
    supportActionBar?.setDisplayShowHomeEnabled(true)
    supportActionBar?.title = title
    presenter.load(genreId)
  }

  override fun onDestroy() {
    super.onDestroy()
    Toothpick.closeScope(this)
  }

  override fun update(data: List<Artist>) {
    adapter.updateData(data)
  }

  override fun onQueueSuccess() {

  }

  override fun onQueueFailure() {

  }

  private fun openProfile(artist: Artist) {
    val intent = Intent(this, ArtistAlbumsActivity::class.java)
    intent.putExtra(ArtistAlbumsActivity.ARTIST_ID, artist.id)
    intent.putExtra(ArtistAlbumsActivity.ARTIST_NAME, artist.name)
    startActivity(intent)
  }

  override fun onMenuItemSelected(menuItem: MenuItem, entry: Artist) {
    val itemId = menuItem.itemId
    when (itemId) {
      R.id.popup_artist_play -> presenter.queue(Queue.NOW, entry)
      R.id.popup_artist_album -> openProfile(entry)
      R.id.popup_artist_queue_next -> presenter.queue(Queue.NEXT, entry)
      R.id.popup_artist_queue_last -> presenter.queue(Queue.LAST, entry)
      R.id.popup_artist_playlist -> {
      }
    }
  }

  override fun onItemClicked(artist: Artist) {
    openProfile(artist)
  }

  companion object {

    const val GENRE_ID = "genre_id"
    const val GENRE_NAME = "genre_name"
  }
}// Required empty public constructor
